### PR TITLE
feat: use JSON RPC for inbound tx observation (including blob tx in Decun upgrade)

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

1. Some necessary code refactor on inbound observation and inbound tracker processing.
2. Uses JSON RPC for inbound tx observation and tracker processing.
3. Added unit tests for above changes.

Closes: [1755](https://github.com/zeta-chain/node/issues/1755)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [x] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [x] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: zetaclient/evm/evm_client.go [hunk 1]
```diff
@@ -1161,13 +1127,6 @@ func (ob *ChainClient) observeERC20Deposited(startBlock, toBlock uint64) uint64
 // returns the last block successfully scanned
 func (ob *ChainClient) observeTssRecvd(startBlock, toBlock uint64, flags observertypes.CrosschainFlags) uint64 {
```
Comment (kevinssgh) [inline zetaclient/evm/evm_client.go:1128]: Is there a reason why we removed the validation of received txns to TSS, for example donations? Also the vote to the core is removed.

File: zetaclient/evm/inbounds.go [hunk 2]
```diff
@@ -59,221 +59,157 @@ func (ob *ChainClient) ObserveTrackerSuggestions() error {
...
 	}
 	return nil
 }
 
-func (ob *ChainClient) CheckReceiptForCoinTypeZeta(txHash string, vote bool) (string, error) {
-	addrConnector, connector, err := ob.GetConnectorContract()
-	if err != nil {
-		return "", err
+// CheckNVoteInboundTokenZeta checks and votes on the given inbound Zeta token
+func (ob *ChainClient) CheckNVoteInboundTokenZeta(tx *ethrpc.Transaction, receipt *ethtypes.Receipt, vote bool) (string, error) {
```
Comment (kevinssgh) [inline zetaclient/evm/inbounds.go:92]: Is the vote flag only used for testing? You can probably use the zetabridge stub I added in my PR #1809

File: zetaclient/evm/evm_client.go [hunk 3]
```diff
@@ -965,10 +942,10 @@ func (ob *ChainClient) observeInTX(sampledLogger zerolog.Logger) error {
 	startBlock, toBlock := ob.calcBlockRangeToScan(confirmedBlockNum, lastScanned, config.MaxBlocksPerPeriod)
 
 	// task 1:  query evm chain for zeta sent logs (read at most 100 blocks in one go)
-	lastScannedZetaSent := ob.observeZetaSent(startBlock, toBlock)
+	lastScannedZetaSent := ob.observeZetaSent(startBlock, toBlock, true)
 
 	// task 2: query evm chain for deposited logs (read at most 100 blocks in one go)
-	lastScannedDeposited := ob.observeERC20Deposited(startBlock, toBlock)
+	lastScannedDeposited := ob.observeERC20Deposited(startBlock, toBlock, true)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/evm_client.go:948]: I was considering future way of unit tests so I added this parameter.

File: zetaclient/evm/inbounds.go (hunk 2 above)
Comment (ws4charlie) [reply zetaclient/evm/inbounds.go:92]: The `false` flag is also used by zetaclientd's `debug` command [these 3 lines](https://github.com/zeta-chain/node/blob/feat-observe-blob-transaction/cmd/zetaclientd/debug.go#L158)

File: zetaclient/evm/evm_client.go (hunk 1 above)
Comment (kevinssgh) [reply zetaclient/evm/evm_client.go:1128]: Thanks, I noticed after.

Comment (lumtis) [inline zetaclient/evm/evm_client.go:1128]: Can we do some renaming here?
```
ObserverTSSReceive
```

File: zetaclient/evm/evm_client.go [hunk 4]
```diff
@@ -1440,7 +1324,7 @@ func (ob *ChainClient) LoadDB(dbPath string, chain common.Chain) error {
 	return nil
 }
 
-func (ob *ChainClient) SetMinAndMaxNonce(trackers []types.OutTxTracker) error {
+func (ob *ChainClient) SetMinAndMaxNonce(trackers []crosschaintypes.OutTxTracker) error {
```
Comment (lumtis) [inline zetaclient/evm/evm_client.go:1327]: Need comment for the function

File: zetaclient/evm/inbounds.go [hunk 5]
```diff
@@ -59,221 +59,157 @@ func (ob *ChainClient) ObserveTrackerSuggestions() error {
 		return err
 	}
```
Comment (lumtis) [inline zetaclient/evm/inbounds.go:60]: I think we can simplify function name
`ObserveTrackerSuggestions` -> `ObserverTrackers` and add comment

File: zetaclient/evm/inbounds.go (hunk 2 above)
Comment (lumtis) [inline zetaclient/evm/inbounds.go:92]: What does `N` means in the function name?

File: zetaclient/evm/utils_test.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+				require.Error(t, err)
+				return
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestCheckEvmTransaction(t *testing.T) {
```
Comment (lumtis) [inline zetaclient/evm/utils_test.go:106]: Why not using table testing here like the function above?

File: zetaclient/evm/constant.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	// https://github.com/zeta-chain/protocol-contracts/blob/d65814debf17648a6c67d757ba03646415842790/contracts/evm/ZetaConnector.base.sol#L54
+	TopicsZetaReverted = 3
+
+	// [signature, recipient, asset]
+	// https://github.com/zeta-chain/protocol-contracts/blob/d65814debf17648a6c67d757ba03646415842790/contracts/evm/ERC20Custody.sol#L43
+	TopicsWithdrawn = 3
+
+	// [signature, asset]
+	// https://github.com/zeta-chain/protocol-contracts/blob/d65814debf17648a6c67d757ba03646415842790/contracts/evm/ERC20Custody.sol#L42
+	TopicsDeposited = 2
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/constant.go:33]: Security wise, the checking on length of evm log may or may not be necessary. I still kept this checking to be safe.

File: zetaclient/evm/evm_client.go [hunk 8]
```diff
@@ -1161,13 +1127,6 @@ func (ob *ChainClient) observeERC20Deposited(startBlock, toBlock uint64) uint64
...
 	if !ob.GetChainParams().IsSupported {
-		//ob.logger.ExternalChainWatcher.Warn().Msgf("observeTssRecvd: chain %d is not supported", ob.chain.ChainId)
-		return startBlock - 1 // lastScanned
-	}
-	// check TSS address (after keygen, ob.Tss.pubkey will be updated)
-	tssAddress := ob.Tss.EVMAddress()
-	if tssAddress == (ethcommon.Address{}) {
-		ob.logger.ExternalChainWatcher.Warn().Msgf("observeTssRecvd: TSS address not set")
 		return startBlock - 1 // lastScanned
 	}
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/evm_client.go:1132]: That's right. There is a `TODO` notation in there. let's track it in here: https://github.com/zeta-chain/node/issues/1847

File: zetaclient/evm/evm_client.go [hunk 9]
```diff
@@ -1185,39 +1144,23 @@ func (ob *ChainClient) observeTssRecvd(startBlock, toBlock uint64, flags observe
...
 			ob.logger.ExternalChainWatcher.Error().Err(err).Msgf("observeTssRecvd: error getting block %d for chain %d", bn, ob.chain.ChainId)
 			return bn - 1 // we have to re-scan from this block next time
 		}
-		if !fallBack {
-			for _, tx := range block.Transactions() {
-				if tx.To() != nil && *tx.To() == tssAddress {
-					if ok := ob.processIntxToTss(tx, bn, block.Hash()); !ok {
-						return bn - 1 // we have to re-scan this block next time
-					}
+		for i := range block.Transactions {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/evm_client.go:1152]: It'll makes the code easy to test. I'll track it with this as a separate refactor: https://github.com/zeta-chain/node/issues/1848

File: zetaclient/evm/inbounds.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/inbounds.go:92]: The common logic they share is the `check confirmations` at the moment. They are respectively dealing with ZetaSent, Deposited and TSS transfer. I think we can leave them as is ATM as these 3 functions are short enough.

File: common/constant.go [hunk 10]
```diff
@@ -3,4 +3,7 @@ package common
 const (
 	// DefaultGasPriceMultiplier is the default gas price multiplier for outbond txs
 	DefaultGasPriceMultiplier = 2
+
+	// DonationMessage is the message for donation transactions
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply common/constant.go:7]: https://github.com/zeta-chain/node/pull/1836/commits/be3e8f9862d885050fb7ac19a3845e62bc8b27cd

File: zetaclient/evm/evm_client.go (hunk 1 above)
Comment (ws4charlie) [reply zetaclient/evm/evm_client.go]: https://github.com/zeta-chain/node/pull/1836/commits/be3e8f9862d885050fb7ac19a3845e62bc8b27cd

File: zetaclient/evm/utils.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
+package evm
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/utils.go]: https://github.com/zeta-chain/node/pull/1836/commits/be3e8f9862d885050fb7ac19a3845e62bc8b27cd

File: zetaclient/evm/evm_client.go (hunk 4 above)
Comment (ws4charlie) [reply zetaclient/evm/evm_client.go]: https://github.com/zeta-chain/node/pull/1836/commits/be3e8f9862d885050fb7ac19a3845e62bc8b27cd

File: zetaclient/evm/inbounds.go (hunk 5 above)
Comment (ws4charlie) [reply zetaclient/evm/inbounds.go:61]: https://github.com/zeta-chain/node/pull/1836/commits/be3e8f9862d885050fb7ac19a3845e62bc8b27cd

File: zetaclient/evm/inbounds.go (hunk 2 above)
Comment (ws4charlie) [reply zetaclient/evm/inbounds.go]: https://github.com/zeta-chain/node/pull/1836/commits/be3e8f9862d885050fb7ac19a3845e62bc8b27cd

File: zetaclient/evm/inbounds.go [hunk 12]
```diff
@@ -387,8 +308,9 @@ func (ob *ChainClient) GetInboundVoteMsgForZetaSentEvent(event *zetaconnector.Ze
 	)
 }
 
-func (ob *ChainClient) GetInboundVoteMsgForTokenSentToTSS(tx *ethtypes.Transaction, sender ethcommon.Address, blockNumber uint64) *types.MsgVoteOnObservedInboundTx {
-	message := hex.EncodeToString(tx.Data())
+// GetInboundVoteMsgForTokenSentToTSS builds a inbound vote message for a token sent to TSS
+func (ob *ChainClient) GetInboundVoteMsgForTokenSentToTSS(tx *ethrpc.Transaction, sender ethcommon.Address, blockNumber uint64) *types.MsgVoteOnObservedInboundTx {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/inbounds.go]: https://github.com/zeta-chain/node/pull/1836/commits/be3e8f9862d885050fb7ac19a3845e62bc8b27cd

File: zetaclient/evm/evm_client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/evm_client.go]: https://github.com/zeta-chain/node/pull/1836/commits/252674798343e2ee3a9caf691ea6334669d9fd4a

File: zetaclient/evm/utils.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+// CheckEvmTxLog checks the basics of an EVM tx log
+func CheckEvmTxLog(vLog *ethtypes.Log, wantAddress ethcommon.Address, wantHash string, wantTopics int) error {
+	if vLog.Removed {
+		return fmt.Errorf("log is removed, chain reorg?")
+	}
+	if vLog.Address != wantAddress {
+		return fmt.Errorf("log emitter address mismatch: want %s got %s", wantAddress.Hex(), vLog.Address.Hex())
+	}
+	if vLog.TxHash.Hex() == "" {
+		return fmt.Errorf("log tx hash is empty: %d %s", vLog.BlockNumber, vLog.TxHash.Hex())
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/utils.go]: https://github.com/zeta-chain/node/pull/1836/commits/252674798343e2ee3a9caf691ea6334669d9fd4a

File: zetaclient/evm/utils.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	"strings"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/onrik/ethrpc"
+	"github.com/pkg/errors"
+)
+
+// CheckEvmTxLog checks the basics of an EVM tx log
+func CheckEvmTxLog(vLog *ethtypes.Log, wantAddress ethcommon.Address, wantHash string, wantTopics int) error {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/utils.go]: https://github.com/zeta-chain/node/pull/1836/commits/252674798343e2ee3a9caf691ea6334669d9fd4a

File: zetaclient/evm/utils.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+	if len(vLog.Topics) != wantTopics {
+		return fmt.Errorf("number of topics mismatch: want %d got %d", wantTopics, len(vLog.Topics))
+	}
+	return nil
+}
+
+// CheckEvmTransaction checks the basics of an EVM transaction
+// Note: these checks are to ensure the transaction is well-formed
+// and can be safely used for further processing by zetaclient
+func CheckEvmTransaction(tx *ethrpc.Transaction) error {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/utils.go]: https://github.com/zeta-chain/node/pull/1836/commits/252674798343e2ee3a9caf691ea6334669d9fd4a

File: zetaclient/evm/inbounds_test.go [hunk 16]
```diff
@@ -1,88 +1,332 @@
 package evm
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/inbounds_test.go]: https://github.com/zeta-chain/node/pull/1836/commits/63836627bc552475f1cffc12b2342b568d7c3192

**Review by kevinssgh — APPROVED** [review]
Looks good

**Comment by ws4charlie** [comment]
> It looks good to me. Wondering if we could consider testing it on local E2E test before moving to production.

It's a good idea. I created a ticket for this as the 1st step: https://github.com/zeta-chain/node/issues/1856

**Pull request merged**